[Property Wrappers] Refactor property wrappers so the synthesized backing init is only type checked once

hborla opened the pull request:
Previously, if a property wrapper had an initial wrapped value expression:

```swift
@Lazy var i = 17
```

Then `PropertyWrapperBackingPropertyInfoRequest` would synthesize and type-check the backing property initializer twice - once with the wrapped value expression, and again with an opaque value placeholder.

Instead, we can have CSApply inject the opaque value placeholder around the `wrappedValue` argument in `coerceCallArguments`.

Resolves: rdar://problem/59685601
Resolves: rdar://problem/56163425
Resolves: [SR-11594](https://bugs.swift.org/browse/SR-11594)

hborla committed:
[NFC] Correct code blocks in property wrapper doc comments

hborla committed:
[Expr] Allow OpaqueValueExpr to store an underlying expression

hborla committed:
[ConstraintSystem] Allow SolutionApplicationTarget to store the
innermost call to init(wrappedValue:) if the target is for initialization
of a property wrapper's backing storage.

hborla committed:
[Property Wrappers] Inject the opaque value placeholder for a property
wrapper original wrapped value expression inside of CSApply.

This prevents type checking the synthesized backing storage initializer
twice - once with the original expression and again with the placeholder.

hborla committed:
[Test] Add a regression test for property wrapper backing initializers that
involve inout closure parameters and were previously type checked twice.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla commented on lib/Sema/CSApply.cpp:
```diff
@@ -298,6 +298,7 @@ namespace {
     ConstraintSystem &cs;
     DeclContext *dc;
     Solution &solution;
+    Optional<SolutionApplicationTarget> target;
```
This shouldn't be `Optional`, but there's a `Solution::coerceCallArguments` that creates an `ExprRewriter` just to call `ExprRewriter::coerceCallArguments`...

DougGregor replied:
We could probably refactor away `Solution::coerceToType` later. It looks unnecessary.

slavapestov commented on include/swift/AST/PropertyWrappers.h:
```diff
@@ -191,11 +191,11 @@ void simple_display(
 
 /// Given the initializer for the given property with an attached property
-/// wrapper, dig out the original initialization expression.
+/// wrapper, dig out the placholder for the original initialization expression.
```
"placeholder"

hborla replied:
Oops, thank you!

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -6359,94 +6361,27 @@ Expr *swift::findOriginalPropertyWrapperInitialValue(VarDecl *var,
-            }
-          }
+      if (auto *opaqueValue = dyn_cast<OpaqueValueExpr>(E)) {
+        if (opaqueValue->getUnderlyingValue() && opaqueValue->isPlaceholder()) {
```
Is it possible for the second condition to be false but the first one true? It looks like you're only setting the underlying value for a placeholder

hborla replied:
No, that's not possible. I only checked both in case something other than property wrappers starts using `underlyingValue` in `OpaqueValueExpr`. It's probably unnecessary, as this code would have to be revised if we start using `underlyingValue` elsewhere anyway

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -5730,13 +5749,35 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
+          auto *placeholder = injectWrappedValuePlaceholder(
+              cs.buildAutoClosureExpr(arg, closureType));
+
+          // We can't pass an opaque closure directly as an autoclosure arg.
```
Is there really a requirement after type checking that the argument to an @autoclosure parameter is syntactically an AutoClosureExpr?

hborla replied:
I think the comment isn't properly explaining why this code is doing all of this wrapping. If the opaque value is a placeholder for an autoclosure (which needs to be the case when `init(wrappedValue:)` takes an escaping autoclosure), we want the affect of autoclosure forwarding here, and the only way to do that is to call the autoclosure when it's passed to the next function (see https://github.com/apple/swift/pull/30537)

hborla commented:
Looks like MovieSwiftUI exposed some bugs. I'll investigate!

hborla committed:
[SILGen] In a memberwise initializer, apply the property wrapper backing
initializer for property wrappers that are not memberwise initialized but
have an explicit original wrapped value.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

DougGregor commented on include/swift/AST/Expr.h:
```diff
@@ -4021,11 +4021,14 @@ class DynamicTypeExpr : public Expr {
 /// subexpressions of that AST node.
 class OpaqueValueExpr : public Expr {
   SourceRange Range;
+  Expr *UnderlyingValue;
```
`OpaqueValueExpr` is a general-purpose construct where the "underlying value" comes from something elsewhere in the AST, above the `OpaqueValueExpr`. Is there some other place we can put the original initializer value that doesn't require adding this field to `OpaqueValueExpr`? One option: create a new `Expr` node `ImplicitlyWrappedPropertyValueExpr`  that stores the `OpaqueValueExpr` and its underlying value. This will make `findWrappedValuePlaceholder` even simpler and avoid altering the design of `OpaqueValueExpr`.

hborla replied:
Yes, I think having a dedicated `Expr` node for the wrapped value is the right thing to do. I was also worried about something else starting to use the `underlyingValue` field that I added here and breaking my assumptions in this PR that it's only used for property wrappers. Thank you!

CodaFi commented on lib/SILGen/SILGenConstructor.cpp:
```diff
@@ -235,27 +250,38 @@ static void emitImplicitValueConstructor(SILGenFunction &SGF,
 
     // If it's memberwise initialized, do so now.
     if (field->isMemberwiseInitialized(/*preferDeclaredProperties=*/false)) {
       FullExpr scope(SGF.Cleanups, field->getParentPatternBinding());
```
I think this scope is subsumed by the one you factored out below.

hborla replied:
You're right, this should be removed. Thank you!

hborla committed:
[Property Wrappers] Introduce a new Expr node for the property wrapper
wrapped value placeholder in an init(wrappedValue:) call that was previously
injected as an OpaqueValueExpr. This commit also restores the old design of
OpaqueValueExpr.

hborla committed:
[SILGen] Remove an unnecessary scope in emitImplicitValueConstructor

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla committed:
[Property Wrappers] Don't continue on to synthesizing and type checking a
property wrapper backing init if the property type doesn't match the
wrapped value type.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test Linux platform

hborla commented:
@swift-ci please smoke test Linux platform

hborla commented:
@swift-ci please smoke test
